crates/proof-of-sql/src/sql/proof_gadgets/shift.rs

```diff
@@ -0,0 +1,117 @@
+    builder: &mut FirstRoundBuilder<'a, S>,
+    num_rows: usize,
+) {
+    builder.produce_rho_evaluation_length(num_rows as usize);
+    builder.request_post_result_challenges(1);
+}
+
+/// Perform final round evaluation of the membership check.
+///
+/// # Panics
+/// Panics if the number of source and candidate columns are not equal.
+#[allow(clippy::too_many_arguments)]
+pub(crate) fn final_round_evaluate_shift<'a, S: Scalar>(
+    builder: &mut FinalRoundBuilder<'a, S>,
+    alloc: &'a Bump,
+    alpha: S,
+    column: Column<'a, S>,
+    shifted_column: Column<'a, S>,
+    rho: Column<'a, S>,
+    ones: &'a [bool],
+    shifted_ones: &'a [bool],
+) {
+    let c_star = alloc.alloc_slice_fill_copy(ones.len(), One::one());
+    column.mul_add(c_star, &alpha);
+    slice_ops::batch_inversion(c_star);
+
+    let d_star = alloc.alloc_slice_fill_copy(shifted_ones.len(), One::one());
+    shifted_column.mul_add(d_star, &alpha);
+    slice_ops::batch_inversion(d_star);
+
+    builder.produce_intermediate_mle(c_star as &[_]);
+    builder.produce_intermediate_mle(d_star as &[_]);
+
+    // sum c_star + c_star * rho - d_star * rho = 0
+    builder.produce_sumcheck_subpolynomial(
+        SumcheckSubpolynomialType::ZeroSum,
+        vec![
+            (S::one(), vec![Box::new(c_star as &[_])]),
+            (S::one(), vec![Box::new(c_star as &[_]), Box::new(rho)]),
+            (-S::one(), vec![Box::new(d_star as &[_]), Box::new(rho)]),
+        ],
+    );
+
+    // c_star + c * c_star - input_ones = 0
+    builder.produce_sumcheck_subpolynomial(
+        SumcheckSubpolynomialType::Identity,
+        vec![
+            (S::one(), vec![Box::new(c_star as &[_])]),
+            (S::one(), vec![Box::new(c_star as &[_]), Box::new(column)]),
+            (-S::one(), vec![Box::new(ones as &[_])]),
+        ],
+    );
+
+    // d_star + d * d_star - shifted_ones = 0
+    builder.produce_sumcheck_subpolynomial(
+        SumcheckSubpolynomialType::Identity,
+        vec![
+            (S::one(), vec![Box::new(d_star as &[_])]),
+            (
+                S::one(),
+                vec![Box::new(d_star as &[_]), Box::new(shifted_column)],
+            ),
+            (-S::one(), vec![Box::new(shifted_ones as &[_])]),
+        ],
+    );
+}
+
+pub(crate) fn verify_shift<S: Scalar>(
+    builder: &mut VerificationBuilder<S>,
+    alpha: S,
+    one_eval: S,
+    shifted_one_eval: S,
+    column_eval: S,
+    candidate_eval: S,
+) -> Result<(), ProofError> {
+    let rho_eval = builder.try_consume_rho_evaluation()?;
+    let c_star_eval = builder.try_consume_final_round_mle_evaluation()?;
+    let d_star_eval = builder.try_consume_final_round_mle_evaluation()?;
+
+    //sum c_star + c_star * rho - d_star * rho = 0
+    builder.try_produce_sumcheck_subpolynomial_evaluation(
+        SumcheckSubpolynomialType::ZeroSum,
+        c_star_eval + c_star_eval * rho_eval - d_star_eval * rho_eval,
+        2,
+    )?;
```

Comment:
Looks like you are doing $\frac{\rho+1}{\alpha+c} \overset{\sum}{=} \frac{\rho}{\alpha+d}$.
Should be

$$\frac{c}{\alpha+\rho+1} \overset{\sum}{=} \frac{d}{\alpha+\rho}$$

Reply:
I think you can simplify a bit more in this case:

$$\hat{c} \cdot (\rho + 1 + \alpha) \equiv c$$
$$\hat{d} \cdot (\rho + \alpha) \equiv d$$
$$\hat{d} \overset{\sum}{=} \hat{c}$$

instead of 


$$(\rho+1)^\star \cdot (\rho + 1 + \alpha) \equiv \chi$$
$$\rho^\star \cdot (\rho + \alpha) \equiv \chi$$
$$d\cdot\rho^\star \overset{\sum}{=} c\cdot (\rho+1)^\star$$